lib/Analysis/Allocation.cpp

```diff
@@ -374,35 +440,81 @@ class AllocationAnalysis {
     });
   }
 
+  void updateBufferRange(BufferT *buffer, IntervalT interval) {
+    if (bufferRange.contains(buffer))
+      interval = interval.merge(bufferRange[buffer]);
+    bufferRange[buffer] = interval;
+    LLVM_DEBUG(llvm::dbgs() << "    UPDATE BUFFER: " << buffer->id << " -> "
+                            << interval << "\n");
+  }
+
   /// Computes the liveness range of the allocated value.
   /// Each buffer is allocated only once.
-  void resolveExplicitBufferLiveness(
-      function_ref<Interval<size_t>(Value value)> getLiveness) {
+  void resolveExplicitBufferLiveness(LivenessF getLiveness) {
     for (auto valueBufferIter : allocation->valueBuffer) {
       auto value = valueBufferIter.first;
       auto *buffer = valueBufferIter.second;
-      bufferRange[buffer] = getLiveness(value);
+      auto ranges = getLiveness(value);
+      // TODO(SJW): bufferRange should support disjoint intervals
+      // This is only a problem for scf loops; cf branches handle
+      // this with aliases
+      updateBufferRange(buffer, ranges.merge());
     }
   }
 
-  /// Extends the liveness range by unionizing the liveness range of the aliased
-  /// values because each allocated buffer could be an alias of others, if block
-  /// arguments are involved.
-  void resolveAliasBufferLiveness(
-      function_ref<Interval<size_t>(Value value)> getLiveness) {
+  void dumpAliasRanges(Value value, const llvm::SetVector<BufferT *> &buffers,
+                       const LiveIntervals &ranges) const {
+    llvm::dbgs() << "    RESOLVE ALIAS: "
+                 << "\n\t";
+    PrintValue(value);
+    llvm::dbgs() << "\n\tRANGES:   ";
+    llvm::for_each(ranges, [cnt = 0](auto interval) mutable {
+      llvm::dbgs() << (cnt++ ? ", " : "") << interval;
+    });
+    llvm::dbgs() << "\n\tBUFFERS:  ";
+    llvm::for_each(buffers, [cnt = 0](auto *buf) mutable {
+      llvm::dbgs() << (cnt++ ? ", " : "") << buf->id;
+    });
+    llvm::dbgs() << "\n";
+  }
+
+  /// Following the alias lattice, an alias that only references a single
```

Comment:
That is correct. In `2`, the reclaimed objects will be Explicit if they are defined by `ttg.local_alloc`. The alias for a loop parameter will extend to all uses within the loop. If it overlaps with the new `ttg.local_alloc`, it cannot reclaim the space. See `cf_loop_carried_overlap` example.

Reply:
My point is that we may want to replace "alias" with a new name because they are not actual aliases. The aliases of an object are those from the same allocation, but in this case they are not. These are objects that reclaim the same memory addresses. Maybe "reuses" could be better name? WDYT

Reply:
OK, I think we could use aliases here. Now I get that you were not trying to say that %gn is also an alias, which is correct.